### PR TITLE
visualization incomplete vis_spec

### LINKED ISSUE #272: Add commented example for creation of visualization files

### PR DESCRIPTION
**Pull request by plakrisenko**
Now visualization should work with 

- empty vis_spec
- vis_spec that contains the only mandatory column plotId
- vis_spec that contains yValues column, but not datasetIds.
In addition to that, two visualization examples are added: about plotting without vis_spec and with vis spec (closes #272)

### CODE REVIEW

File: petab/visualize/helper_functions.py [hunk 1]
```diff
@@ -390,56 +396,197 @@ def get_default_vis_specs(
     return vis_spec, exp_data
 
 
-def check_ex_visu_columns(vis_spec: pd.DataFrame,
-                          dataset_id_list: List[IdsList],
-                          legend_dict: Dict) -> pd.DataFrame:
+def get_vis_spec_dependent_columns_dict(
```
Comment (dweindl) [inline petab/visualize/helper_functions.py]: Codacy Issue found: [Missing function or method docstring](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=5263076)

File: petab/visualize/helper_functions.py [hunk 2]
```diff
@@ -351,9 +356,10 @@ def get_default_vis_specs(
     if group_by != 'dataset':
         # datasetId_list will be created (possibly overwriting previous list
         #  - only in the local variable, not in the tsv-file)
-        exp_data, dataset_id_list, legend_dict = create_dataset_id_list(
-            sim_cond_id_list, sim_cond_num_list, observable_id_list,
-            observable_num_list, exp_data, exp_conditions, group_by)
+        exp_data, dataset_id_list, legend_dict, yvalues_dict = \
```
Comment (dweindl) [inline petab/visualize/helper_functions.py]: Codacy Issue found: [Unused variable 'yvalues_dict'](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=5263076)

File: petab/visualize/helper_functions.py [hunk 3]
```diff
@@ -390,56 +396,197 @@ def get_default_vis_specs(
...
+                         X_VALUES, X_OFFSET, X_LABEL, X_SCALE, Y_OFFSET,
+                         Y_LABEL, Y_SCALE, LEGEND_ENTRY]
+
+    for column in vis_spec.columns:
+        if column in columns_to_expand:
+            column_entries = []
+            if Y_VALUES in vis_spec.columns:
+                for i, plot_id in enumerate(columns_dict[PLOT_ID]):
+                    select_conditions = (vis_spec[PLOT_ID] == plot_id) & (
+                            vis_spec[Y_VALUES] == columns_dict[Y_VALUES][i])
```
Comment (dweindl) [inline petab/visualize/helper_functions.py]: Codacy Issue found: [Wrong hanging indentation (remove 4 spaces).](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=5263076)

File: petab/visualize/helper_functions.py [hunk 4]
```diff
@@ -390,56 +396,207 @@ def get_default_vis_specs(
...
```
Comment (dweindl) [inline petab/visualize/helper_functions.py]: Codacy Issue found: [Wrong hanging indentation (remove 4 spaces).](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=5263076)

File: petab/visualize/helper_functions.py [hunk 5]
```diff
@@ -390,56 +396,208 @@ def get_default_vis_specs(
...
+        vis_spec = pd.DataFrame(columns_dict)
+    else:
+        # TODO: do validation
+        # so, plotid is definitely there
+        if DATASET_ID not in vis_spec.columns:
+            if Y_VALUES in vis_spec.columns:
+                plot_id_list = np.unique(vis_spec[PLOT_ID])
+                observable_id_list = [
+                    vis_spec[vis_spec[PLOT_ID] ==
+                             plot_id].loc[:, Y_VALUES].values
```
Comment (dweindl) [inline petab/visualize/helper_functions.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/dweindl/PEtab/pullRequest?prid=5263076)

File: petab/visualize/helper_functions.py [hunk 6]
```diff
@@ -390,56 +396,208 @@ def get_default_vis_specs(
...
+                                                exp_conditions,
+                                                dataset_id_list,
+                                                sim_cond_id_list,
+                                                sim_cond_num_list,
+                                                observable_id_list,
+                                                observable_num_list)
+        vis_spec = pd.DataFrame(columns_dict)
+    else:
+        # TODO: do validation
+        # so, plotid is definitely there
```
Comment (MerktSimon) [inline petab/visualize/helper_functions.py]: validation meaning something like `PLOT_ID in vis_spec.columns`?

File: petab/visualize/helper_functions.py [hunk 7]
```diff
@@ -390,56 +396,208 @@ def get_default_vis_specs(
...
+                                      for plot_id in plot_id_list]
+                exp_data, columns_dict = \
+                    get_vis_spec_dependent_columns_dict(
+                        exp_data,
+                        exp_conditions,
+                        observable_id_list=observable_id_list)
+            else:
+                # PLOT_ID is there, but NOT DATASET_ID and not Y_VALUES,
+                # but potentially some settings.
+                # TODO: multiple plotids with diff settings
```
Comment (MerktSimon) [inline petab/visualize/helper_functions.py]: this does not work with multiple plotIDs right now?
If it does not raise warning. If it does remove comment

**Comment by MerktSimon** [comment]
and documentation_data_format.md is has still the old mandatory columns

File: petab/core.py [hunk 8]
```diff
@@ -46,7 +46,11 @@ def get_visualization_df(visualization_file: str) -> pd.DataFrame:
     Returns:
         Visualization DataFrame
     """
-    return pd.read_csv(visualization_file, sep="\t", index_col=None)
+    try:
+        vis_spec = pd.read_csv(visualization_file, sep="\t", index_col=None)
+    except pd.errors.EmptyDataError:
+        vis_spec = pd.DataFrame()
+    return vis_spec
```
Comment (MerktSimon) [inline petab/core.py]: maybe include warning? Usually users probably do not intend to have empty visu files

**Comment by MerktSimon** [comment]
- "If no settings are provided, the measurements will be plotted with default settings. Namely, it will be assumed that the independent variable is time and **....** "  in https://github.com/PEtab-dev/PEtab/blob/pl_visualization/doc/example/example_visualization_with_visspec.ipynb is missing a something.

- Todo in "If you want to have more control over what will be plotted, refer to this example TODO" in https://github.com/PEtab-dev/PEtab/blob/pl_visualization/doc/example/example_visualization_without_visspec.ipynb

- "TODO: plotted_noise TODO: simulations plot_data_and_simulation" at the end of https://github.com/PEtab-dev/PEtab/blob/pl_visualization/doc/example/example_visualization_without_visspec.ipynb

Personally I would put todos in an issue and remove them from example but maybe that is just a matter of taste

File: petab/visualize/helper_functions.py [hunk 9]
```diff
@@ -390,56 +396,208 @@ def get_default_vis_specs(
...
-    if X_LABEL not in vis_spec.columns:
-        vis_spec[X_LABEL] = 'time'
+        vis_spec[X_VALUES] = 'time'
     if X_OFFSET not in vis_spec.columns:
         vis_spec[X_OFFSET] = 0
+    if X_LABEL not in vis_spec.columns:
+        vis_spec[X_LABEL] = 'time'
+    if X_SCALE not in vis_spec.columns:
+        vis_spec[X_SCALE] = LIN
+    # TODO: Y_VALUES column will already be there?
```
[the remark this answers is not in the document]
Comment (plakrisenko) [reply petab/visualize/helper_functions.py]: I meant that maybe the code was reorganized in such a way that Y_VALUES will always be in vis_spec when `check_ex_visu_columns` is called. But now i think, that lines 575, 576 are still needed for the case when DATASET_ID is in vis_spec file but not Y_VALUES.

File: petab/visualize/helper_functions.py (hunk 6 above)
Comment (plakrisenko) [reply petab/visualize/helper_functions.py]: Yes and #190

File: petab/visualize/helper_functions.py (hunk 7 above)
Comment (plakrisenko) [reply petab/visualize/helper_functions.py]: vis_spec like this won't work right now

plotId | xOffset | yScale
-- | -- | --
plot1 | 100 | log
plot2 | 50 | log
plot3 | 0 | lin

I don't know if anyone would want to use such vis_spec. Potentially could be tried to plot everything with different settings.

File: petab/visualize/helper_functions.py (hunk 6 above)
Comment (MerktSimon) [reply petab/visualize/helper_functions.py]: ah I see. you could link this issue in the comment

**Pull request merged**